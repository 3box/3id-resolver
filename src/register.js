import { verifyJWT } from 'did-jwt'
import DidDocument from 'ipfs-did-document'
import base64url from 'base64url'
import { registerMethod } from 'did-resolver'

const PUBKEY_IDS = ['signingKey', 'managementKey', 'encryptionKey']
const SUB_PUBKEY_IDS = ['subSigningKey', 'subEncryptionKey']

function register (ipfs, opts = {}) {
  registerMethod('3', (_, { id }) => resolve(ipfs, id))
}

async function resolve (ipfs, cid, isRoot) {
  let doc
  try {
    doc = await DidDocument.cidToDocument(ipfs, cid)
    validateDoc(doc)
    if (doc.root) {
      if (isRoot) throw new Error('Only one layer subDoc allowed')
      const rootDoc = await resolve(ipfs, doc.root.split(':')[2], true)
      await verifyProof(doc)
      doc = mergeDocuments(rootDoc, doc)
    }
  } catch (e) {
    try {
      await ipfs.pin.rm(cid)
    } catch (e) {}
    throw new Error('Invalid 3ID')
  }
  return doc
}

function validateDoc (doc) {
  let pubKeyIds = PUBKEY_IDS
  if (!doc || !doc.publicKey || !doc.authentication) {
    throw new Error('Not a valid 3ID')
  }
  if (doc.root) {
    pubKeyIds = SUB_PUBKEY_IDS
    if (!doc.space) throw new Error('Not a valid 3ID')
  }
  doc.publicKey.map(entry => {
    const id = entry.id.split('#')[1]
    if (!pubKeyIds.includes(id)) throw new Error('Not a valid 3ID')
  })
}

function encodeSection (data) {
  return base64url.encode(JSON.stringify(data))
}

async function verifyProof (subDoc) {
  const subSigningKey = subDoc.publicKey.find(entry => entry.id.includes(SUB_PUBKEY_IDS[0])).publicKeyHex
  const subEncryptionKey = subDoc.publicKey.find(entry => entry.id.includes(SUB_PUBKEY_IDS[1])).publicKeyBase64
  const payload = encodeSection({
    iat: null,
    subSigningKey,
    subEncryptionKey,
    space: subDoc.space,
    iss: subDoc.root
  })
  const header = encodeSection({ typ: 'JWT', alg: subDoc.proof.alg })
  const jwt = `${header}.${payload}.${subDoc.proof.signature}`
  await verifyJWT(jwt)
}

function mergeDocuments (doc, subDoc) {
  subDoc.publicKey = doc.publicKey.concat(subDoc.publicKey)
  return subDoc
}

module.exports = register
