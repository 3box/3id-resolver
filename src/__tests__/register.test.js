import resolve from 'did-resolver'
import base64url from 'base64url'
import register from '../register'
import DidDocument from 'ipfs-did-document'
import { SimpleSigner, createJWT } from 'did-jwt'
import IPFS from 'ipfs'

const PRIV_KEY = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const PUB_KEY = '03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479'


describe('3ID Resolver', () => {
  let ipfs, ipfsd

  beforeAll(async () => {
    ipfs = await initIPFS()
    register(ipfs)
  })

  afterAll(async () => {
    await ipfs.stop()
  })

  describe('resolve 3ID', () => {
    let rootDID

    describe('root 3ID', async () => {
      let doc

      beforeEach(async () => {
        doc = new DidDocument(ipfs, '3')
      })

      it('throws on invalid document', async () => {
        doc.addPublicKey('signingKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', 'not a pub key')
        await doc.commit({ noTimestamp: true })
        await expect(resolve(doc.DID)).rejects.toEqual(new Error('Invalid 3ID'))
      })

      it('resolves valid document', async () => {
        doc.addPublicKey('signingKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', PUB_KEY)
        doc.addPublicKey('encryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', 'fake encryptionKey')
        doc.addPublicKey('managementKey', 'Secp256k1VerificationKey2018', 'ethereumAddress', 'fake eth addr')
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'signingKey')
        const cid = await doc.commit({ noTimestamp: true })
        const rawDoc = await DidDocument.cidToDocument(ipfs, cid)
        await expect(resolve(doc.DID)).resolves.toEqual(rawDoc)
        rootDID = doc.DID
      })
    })

    describe('sub 3ID', () => {
      let doc

      beforeEach(async () => {
        doc = new DidDocument(ipfs, '3')
      })

      it('throws on invalid proof', async () => {
        doc.addPublicKey('subSigningKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', 'fake subSigningKey')
        doc.addPublicKey('subEncryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', 'fake subEncryptionKey')
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'subSigningKey')
        doc.addCustomProperty('space', 'a space name')
        doc.addCustomProperty('root', rootDID)
        const signature = base64url.encode('obviously invalid signature')
        doc.addCustomProperty('proof', { alg: 'ES256K', signature })
        await doc.commit({ noTimestamp: true })

        await expect(resolve(doc.DID)).rejects.toEqual(new Error('Invalid 3ID'))
      })

      it('resolves on valid sub document', async () => {
        const subSigningKey = 'fake subSigningKey'
        const subEncryptionKey = 'fake subEncryptionKey'
        const space = 'a space name'
        doc.addPublicKey('subSigningKey', 'Secp256k1VerificationKey2018', 'publicKeyHex', subSigningKey)
        doc.addPublicKey('subEncryptionKey', 'Curve25519EncryptionPublicKey', 'publicKeyBase64', subEncryptionKey)
        doc.addAuthentication('Secp256k1SignatureAuthentication2018', 'subSigningKey')
        doc.addCustomProperty('space', space)
        doc.addCustomProperty('root', rootDID)
        const payload = {
          subSigningKey,
          subEncryptionKey,
          space,
          iat: null
        }
        const signer = new SimpleSigner(PRIV_KEY)
        const jwt = await createJWT(payload, {
          issuer: rootDID,
          signer,
          alg: 'ES256K'
        })
        const signature = jwt.split('.')[2]
        doc.addCustomProperty('proof', { alg: 'ES256K', signature })
        await doc.commit({ noTimestamp: true })

        await expect(resolve(doc.DID)).resolves.toMatchSnapshot()
      })
    })
  })
})

const initIPFS = async () => {
  return new Promise((resolve, reject) => {
    let ipfs = new IPFS({
      repo: './.ipfs',
      config: { Addresses: {}, Discovery: {}, Bootstrap: [] }
    })
    ipfs.on('error', reject)
    ipfs.on('ready', () => resolve(ipfs))
  })
}

      //const pubkeys = this._keyrings[spaceName].getPublicKeys()
    //}
