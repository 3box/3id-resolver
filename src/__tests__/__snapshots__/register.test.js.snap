// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3ID Resolver resolve 3ID sub 3ID resolves on valid sub document 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:zdpuAsUryAVEbMHNi4Xdm8AYSohJoMKgqc9mbQDJEeGPVTPG2#subSigningKey",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:zdpuAsUryAVEbMHNi4Xdm8AYSohJoMKgqc9mbQDJEeGPVTPG2",
  "proof": Object {
    "alg": "ES256K",
    "signature": "rF8U6R9Tzt1WW_AFlXX69ZOQvb8cd9m0fPAw3A9jmDST2-eikw02sstDEINP7M1ZQL4rho1Rv-zXEtMDajf-bQ",
  },
  "publicKey": Array [
    Object {
      "id": "did:3:zdpuAtgNc4w55mZTYL2YLhDNZgf2mrLnJ8k2Vt4Pccrxbqh6n#signingKey",
      "publicKeyHex": "03fdd57adec3d438ea237fe46b33ee1e016eda6b585c3e27ea66686c2ea5358479",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:zdpuAtgNc4w55mZTYL2YLhDNZgf2mrLnJ8k2Vt4Pccrxbqh6n#encryptionKey",
      "publicKeyBase64": "fake encryptionKey",
      "type": "Curve25519EncryptionPublicKey",
    },
    Object {
      "ethereumAddress": "fake eth addr",
      "id": "did:3:zdpuAtgNc4w55mZTYL2YLhDNZgf2mrLnJ8k2Vt4Pccrxbqh6n#managementKey",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:zdpuAsUryAVEbMHNi4Xdm8AYSohJoMKgqc9mbQDJEeGPVTPG2#subSigningKey",
      "publicKeyHex": "fake subSigningKey",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "id": "did:3:zdpuAsUryAVEbMHNi4Xdm8AYSohJoMKgqc9mbQDJEeGPVTPG2#subEncryptionKey",
      "publicKeyBase64": "fake subEncryptionKey",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
  "root": "did:3:zdpuAtgNc4w55mZTYL2YLhDNZgf2mrLnJ8k2Vt4Pccrxbqh6n",
  "space": "a space name",
}
`;
